### ISSUE TITLE
Build this repo or nuget packate

### ISSUE THREAD

**Issue by Defkon1** [ComponentFactory/Krypton#56]
Is it better to download/fork/build this project and add the DLLs to my project or to add KS from a nuget packages like https://www.nuget.org/packages/KryptonSuite47/ ?

**Comment by JoeRecursionJoe** [comment]
I tried both Nuget packages and received a security error something like "attempt by security transparent method componentfactory.krypton to call native code through method...". 

The Nuget Krypton dlls version are 4.7 and this repo dll version is 4.5 so I'm not sure if someone forked it and modified it to get the Nuget.

I would stay with this repo as it comes from the author...

**Comment by Wagnerp** [comment]
As the author of the Nuget packages, I would suggest to download/build and test the original repo. **There is a odd bug with the Nuget packages, whereby it will not work unless you have the original DLLs working.** This is something I cannot resolve, as it has something to do with how the original author structured the code. 

However, it does work, as I had fully tested the new DLLs before publishing. The only change I had made was to re-target the frameworks to 4.6.2 and 4.7. I also added the DPI awareness feature in version 4.7, though I'm not sure that this is working. Hope this helps.

**Comment by JoeRecursionJoe** [comment]
Thanks. Could you elaborate though on , “it will not work unless you have the original dlls working” because I had the original dlls from this repo installed, but still had the Nuget the error. 

Could the Nuget error possibly be related to changing the frameworks to 4.6.2 and 4.7?

**Comment by Wagnerp** [comment]
Hi @JoeRecursionJoe 

Thanks for your reply. Yes I can explain why it happens, but I need to write it up first, as it's quite long. The reason for the disparity in version numbers is an easy reference to the frameworks that they were compiled for. The short answer is yes they do work, but you need to put in some extra work to get them working. I will post a more detailed explanation on this thread either later today or tomorrow.

**Comment by Wagnerp** [comment]
Hi @Defkon1 & @JoeRecursionJoe 

After a bit of testing, I have discovered that the DLLs downloaded by the Nuget packages may need to be re-registered into your Global Assembly Cache. These DLLs can also be temperamental sometimes.

To do this, simply build your project, go to your <directory name>\bin folder, copy the krypton DLLs into another folder i.e. C:\Developer DLLs\Krypton Suite\Versions\<version number>. Find a copy of gacutil.exe on your computer, copy/paste it into the mentioned folder (you may need to download/install the latest ‘’stable’’ Windows SDK to find this binary).

Open up notepad and paste the following into it:-
```
gacutil.exe -i "ComponentFactory.Krypton.Design.dll"
gacutil.exe -i "ComponentFactory.Krypton.Docking.dll"
gacutil.exe -i "ComponentFactory.Krypton.Navigator.dll"
gacutil.exe -i "ComponentFactory.Krypton.Ribbon.dll"
gacutil.exe -i "ComponentFactory.Krypton.Toolkit.dll"
gacutil.exe -i "ComponentFactory.Krypton.Workspace.dll"

::REBOOT:
shutdown.exe /r /t 00
```

Save the file as ‘’RegisterToGAC.bat’’ and place it into the directory where you have stored the DLLs and gacutil.exe. Next open a new command prompt window with administrative privileges, navigate to the directory you have stored your files, and then run ‘’RegisterToGAC.bat’’.

‘’PLEASE ENSURE YOU HAVE SAVED ANY IMPORTANT WORK BEFORE RUNNING THIS BATCH FILE, AS YOUR SYSTEM WILL REBOOT AND I WILL NOT BE RESPONSIBLE FOR ANY LOSS OF DATA. YOU DO THIS AT YOUR OWN RISK!!!’’

‘’NOTE:’’ It may take several reboots to register the DLLs into your Global Assembly Cache!
Open your project, and add the DLLs to your toolbox. Hopefully this should work! On several occasions, I have noticed a Heisenbug effect.

Alternatively, you can download the original repo, and re-target frameworks/rebuild/re-register yourself.

Please note that the 4.6.2 version will run on Windows 7 and above, whereas version 4.7 will ‘’ONLY’’ run on Windows 10 Creators Update (1703) and above!

I hope this helps!

**Comment by JoeRecursionJoe** [comment]
Great @Wagnerp ! I am assuming you would only have to register to GAC only once?

Also I've never created a Nuget package, where do you go to find out how?

**Comment by Wagnerp** [comment]
Hi @JoeRecursionJoe 

I have found that sometimes you may need to run the script a few times before it finally 'clicks' into place, or even weirder, you can only use it after 12 hours of running the script, and shutting down the system before booting it up again.

As for creating Nuget packages, I use a tool called **NuGetPackageExplorer** available from [https://github.com/NuGetPackageExplorer/NuGetPackageExplorer](url)

**Comment by JoeRecursionJoe** [comment]
Cool thanks @Wagnerp I'll check it out...

**Comment by Wagnerp** [comment]
@JoeRecursionJoe If it doesn't work out for you, then I would download the original repo, and re-target frameworks/rebuild/re-register the binaries.

**Comment by FreeFlyingEnchilada** [comment]
I'd recommend downloading the repo. It builds with 0 modifications in Visual Studio 2017. The one issue is that you need to run the instance of visual studio as an administrator since the build automatically adds the generated DLLs to the global assembly cache using gacutil.exe, which requires elevation.

More importantly, the repo contains extensive examples and documentation you would not get with just a nuget of the dlls. The code itself is also very well documented and also works as a resource.

**Comment by Zyxious** [comment]
I also had problem using the nuget packages in VS.NET 2017 with runtime 4.6.2. Using the error message I think I figure out the problem. 

The Krypton suite is developed for .NET 2 and not adapted to .NET 4. The code access security has been changed between the .NET versions. Red more on this here: <https://docs.microsoft.com/en-us/dotnet/framework/misc/security-transparent-code-level-1>

I'm able to use the Krypton Suite now without register the dll:s to the GAC by:

- Changing the target framework on all component projects to 4.6.2
- Adding the code below to all component projects AssemblyInfo.cs file
  [assembly: SecurityRules(SecurityRuleSet.Level1)] 
- Compile all component projects and import the resulting dll:s into my .NET 4.6.2 WinForms project

@Wagnerp, maybe you could do these changes and update the nuget packages?

**Comment by Wagnerp** [comment]
@Zyxious 

Of course I'll see what I can do. Is it just adding the [assembly: SecurityRules(SecurityRuleSet.Level1)] into all AssemblyInfo.cs files or is there more to do? I have got version 4.7 working in my environment, I may remove that package, as everyone seems to have issues with it.

**Comment by Zyxious** [comment]
Yes, I opened the Krypton Components\Krypton Components Suite 2015.sln.
Changed target framework in all projects of the solution to (in my case) .NET 4.6.2. and I also added [assembly: SecurityRules(SecurityRuleSet.Level1)] to the assemblyinfo.cs file in each project.

I have done no more modifications.

/Andy


[cid:image001.png@01D31D83.BD5B39B0]



Från: Wagnerp [mailto:notifications@github.com]
Skickat: den 25 augusti 2017 09:13
Till: ComponentFactory/Krypton <Krypton@noreply.github.com>
Kopia: Yu, Andy <Andy.Yu@electra.se>; Mention <mention@noreply.github.com>
Ämne: Re: [ComponentFactory/Krypton] Build this repo or nuget packate (#56)


@Zyxious<https://github.com/zyxious>

Of course I'll see what I can do. Is it just adding the [assembly: SecurityRules(SecurityRuleSet.Level1)] into all AssemblyInfo.cs files or is there more to do? I have got version 4.7 working in my environment, I may remove that package, as everyone seems to have issues with it.

**Comment by Wagnerp** [comment]
Hi Andy (@Zyxious),

Thanks for the reply, I will work on it today. Please note that the updated section of my previous comment is more aimed at the readers of this thread and not directly aimed at you.

Peter

**Comment by Wagnerp** [comment]
I've just created a new version of the Nuget package which will be uploaded shortly. This includes the fixes as suggested by @Zyxious. I've also included some other helpful files, which will be added to a new folder labeled 'Extras' in your project's bin directory. As a result, the file is now 66 MB with these extra files.

I hope to upload a .NET 4.7 version soon.

**Comment by Smurf-IV** [comment]
Is this the new download ?
[https://www.nuget.org/packages/KryptonSuite47b/](https://www.nuget.org/packages/KryptonSuite47b/)

**Comment by Wagnerp** [comment]
Hi @Smurf-IV 

Yes it is

**Comment by Smurf-IV** [comment]
It appears that the designer dll does not work within Studio, This may be because you have not updated the version number usage in the actual implementation dll's e.g.
If the designer dll has 
[assembly: AssemblyVersion("**4.7.1.0**")]

then the KryptonButton class will **need** to have it's attribute line changed to 

[Designer("ComponentFactory.Krypton.Toolkit.KryptonButtonDesigner, ComponentFactory.Krypton.Design, Version=**4.7.1.0**, Culture=neutral, PublicKeyToken=a87e673e9ecb6e8e")]

**Comment by Smurf-IV** [comment]
@Wagnerp, where is your updated fork for the .Net 4.7 codebase. I have the same done, but with .net47 syntactic sugar etc. added (i.e. I've let ReSharper loose on the code base). I would like to combine and then for you to review and nuget it out.

**Comment by Wagnerp** [comment]
@Smurf-IV Will upload it later today. Please check back later on my profile to submit a merge.

**Comment by oghenez** [comment]
Can i recompile this for .Net 2.0?

**Comment by Wagnerp** [comment]
@oghenez 

Yes you can, just change the target framework back to .NET 2.0.
